### PR TITLE
Split common vagrant tasks into base class DiskFormatVagrantBase

### PR DESCRIPTION
**Pull request by dcermak**
Fixes part of the issues addressed in #947.

Changes proposed in this pull request:
* Split `DiskFormatVagrantLibVirt` into `DiskFormatVagrantLibVirt` and `DiskFormatVagrantBase`, as creating a vagrant box involves a few steps that have to be done independently of the provider. The base class performs these instead of the libvirt specific class.
* Split the unit test for `DiskFormatVagrantLibVirt` into tests for the base class (which I have tried to write so that they test the documented behavior of the base class) and test of the libvirt class
* I have modified the `Vagrantfile` that is bundled with the libvirt provider to match the one that packer produces.

### CODE REVIEW

File: kiwi/storage/subformat/vagrant_libvirt.py [hunk 1]
```diff
@@ -83,71 +44,24 @@ def create_image_format(self):
...
-            random.randrange(0, 0xff)
-        )
+    def additional_vagrant_settings(self):
+        # () -> str:
+        """
+        Returns settings for the libvirt provider telling vagrant to use kvm.
+        """
+        return """config.vm.provider :libvirt do |libvirt|
+    libvirt.driver = "kvm"
+  end"""
```
Comment (schaefi) [inline kiwi/storage/subformat/vagrant_libvirt.py]: Could you use the dedent module for returning strings like this:

```python
from textwrap import dedent

libvirt_driver_setup = dedent('''
    config.vm.provider :libvirt do |libvirt|
        libvirt.driver = "kvm"
    end
''')

return libvirt_driver_setup
```

This is imho nicer to read and explicit in what it returns

Comment (dcermak) [reply kiwi/storage/subformat/vagrant_libvirt.py]: I'll give it a try, but I think I didn't use it because it broke the indentation of the resulting `Vagrantfile` that ends up in the box.

Comment (schaefi) [reply kiwi/storage/subformat/vagrant_libvirt.py]: The dedent module allows to provide the indentation as you need it. If you strip() and add a linesep at the end it should be identical to the version you had in """ quotes

Comment (dcermak) [reply kiwi/storage/subformat/vagrant_libvirt.py]: I have changed the way `_create_box_vagrantconfig` pastes the customized vagrant config and the result looks better now, albeit requiring another call to `indent`.

Comment (schaefi) [reply kiwi/storage/subformat/vagrant_libvirt.py]: from textwrap import dedent

that exists in py2 and py3, kiwi already uses that module for years :)

Comment (schaefi) [reply kiwi/storage/subformat/vagrant_libvirt.py]: Still like the version from above more. Sorry but

```
 "  " + "  ".join(
            dedent("""
```

looks strange to me. Also python uses single quotes for text that is taken one by one.

```python
dedent('''
    config.vm.provider :libvirt do |libvirt|
        libvirt.driver = "kvm"
    end
''').strip()
```

produces the same formatting result and avoids and leading spaces. Any reason why that does not work for you ?

Comment (schaefi) [reply kiwi/storage/subformat/vagrant_libvirt.py]: Also saw you are using splitlines() now which turns the string including espace sequences into a list. That invalidates the doc string because the method now returns a list and not a string anymore. Also why is this better than before ? It's starting to confuse me.

Comment (schaefi) [reply kiwi/storage/subformat/vagrant_libvirt.py]: The concept of additional_vagrant_settings() is ok for me as it just added raw data to the vagrantfile. Maybe the method should be better named additional_vagrant_raw_config() or similar. If you wish to add more structure I suggest to pass a dictionary with information, but in case of the Vagrantfile which is actually ruby code the idea to pass raw data is probably the more lightweight solution. So I'm ok with the string but the list I don't understand

Comment (dcermak) [reply kiwi/storage/subformat/vagrant_libvirt.py]: Yes, I actually need the two leading spaces, otherwise it'll mess up the `Vagrantfile`.

Since it is quite confusing having to do that in child classes, I have moved the indentation code to `DiskFormatVagrantBase._create_box_vagrantconfig`, so that this is done once for all children.

Comment (schaefi) [reply kiwi/storage/subformat/vagrant_libvirt.py]: ok, thanks for the changes. I was reading the code again and I think we should do the vagrantconfig handling a bit different. My idea is to use a simple template. I hope you don't mind but that spacing and such could bite us in the future and is not very flexible. So here is my idea

```python
import os
  
from textwrap import dedent
from string import Template

class VagrantConfigTemplate(object):
    """
    **vagrant config template**
    """
    def __init__(self):
        self.indent = '  ';

        self.header = dedent('''
            Vagrant.configure("2") do |config|
        ''').strip() + os.linesep

        self.mac = self.indent + dedent('''
            config.vm.base_mac = ${mac_address}
        ''').strip() + os.linesep

        self.end = dedent('''
            end
        ''').strip()

    def get_template(self, custom_settings=None):
        template_data = self.header
        template_data += self.mac
        if custom_settings:
            template_data += self.indent
            template_data += self.indent.join(
                custom_settings.splitlines(True)
            )
            template_data += os.linesep
        template_data += self.end
        return Template(template_data)

def get_additional_vagrant_config_settings():
    return dedent('''
        config.vm.provider :libvirt do |libvirt|
          libvirt.driver = "kvm"
        end
    ''').strip()

def _create_box_vagrantconfig():
    template = VagrantConfigTemplate()
    vagrant_config = template.get_template(
        custom_settings=get_additional_vagrant_config_settings()
    )
    return vagrant_config.substitute(
        {'mac_address': 'foo'}
    )

print(_create_box_vagrantconfig())
```

What do you think ? It will connect all indention and layouting into a template class and also allows for better testing.

Thanks for your patience

Comment (dcermak) [reply kiwi/storage/subformat/vagrant_libvirt.py]: I'll give your idea a try and code something up so that we find out its pain points. I agree that just pasting stuff into the `Vagrantfile` (as I have done it) is pretty ugly and could result in bad things happening.

Comment (schaefi) [reply kiwi/storage/subformat/vagrant_libvirt.py]: great, thanks :+1:

Comment (dcermak) [reply kiwi/storage/subformat/vagrant_libvirt.py]: Done. I have refactored the code a little bit and converted it to a single function. Since I added a doctest to this function I tried to enable them for pytest, unfortunately it does not seem to collect them properly.

File: kiwi/storage/subformat/vagrant_base.py [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+    template_data = HEADER
+    template_data += MAC
+    if custom_settings:
+        template_data += INDENT
+        template_data += INDENT.join(
+            custom_settings.splitlines(True)
+        )
+        template_data += os.linesep
+    template_data += END
+    return Template(template_data)
```
Comment (schaefi) [inline kiwi/storage/subformat/vagrant_base.py]: Could we please put the template into its own class. We use this concept in kiwi in other places too. See e.g ```kiwi/storage/subformat/template/vmware_settings.py``` or ```kiwi/bootloader/template/```

File: kiwi/storage/subformat/vagrant_base.py [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+    end
+
+    If your provider/box requires additional settings, provide them as follows:
+    >>> extra_settings = dedent('''
+    ... config.vm.hostname = "no-dead-beef"
+    ... config.vm.provider :special do |special|
+    ...   special.secret_settings = "please_work"
+    ... end
+    ... ''').strip()
+    >>> print(get_vagrantfile_template(extra_settings).substitute({'mac_address': 'DEADBEEF'}))
```
Comment (schaefi) [inline kiwi/storage/subformat/vagrant_base.py]: That information imho belongs to a template class specification. For coding examples we use

```python
"""
.. code:: python
     ...
"""
```

The :returns: as well as :Example: as well as :type: seems to be not evaulated by the sphinx docs. So for method doc strings we use this format all over the place

```
def foo(arg):
    """
    Does something

    :param type arg: arg for foo

    :return: whatever it returns

    :rtype: return_type
    """
```

Could you adapt that to be consistent with the rest of the code base ? I know there are many ways to do this but it wold be good to have a common schema

File: kiwi/storage/subformat/vagrant_base.py [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+            ),
+            use_for_bundle=True,
+            compress=False,
+            shasum=True
+        )
+
+    @classmethod
+    def get_additional_metadata(cls):
+        # type: () -> Dict[str, str]
+        return {}
```
Comment (schaefi) [inline kiwi/storage/subformat/vagrant_base.py]: Could we avoid that type of comments: In this particular method it's also wrong. it returns an empty dict
methods of the public interface should have a doc string as explained before. private methods (_foo) doesn't need it and could take some information if not clear by the code. I usually avoid any comment block that is of no interest because not in the public api and try to write the code to be as readable as possible. From my experience if there are comment blocks that looks like code either the comment or the code is wrong :) So let's focus on the minimal needed and be as explicit and clear in the code as possible. Hope that makes sense to you too

File: kiwi/storage/subformat/vagrant_base.py [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+
+    @classmethod
+    def get_additional_vagrant_config_settings(cls):
+        # type: () -> str
+        return ""
```
Comment (schaefi) [inline kiwi/storage/subformat/vagrant_base.py]: It should just pass and avoid that comment block. When using a template class there is no need for the method to return an empty string. But it should mention that this method can be overwritten by the specialized provider
class. e.g

```python
@classmethod
def get_additional_vagrant_config_settings(cls):
    """
    Handle additional configurations in the Vagrantfile

    Depending on the provider additional vagrant configuration settings are required.
    The implementation of those are done in the specialized provider class.
    """
    pass
```

File: kiwi/storage/subformat/vagrant_base.py [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+        vagrant_config = get_vagrantfile_template(
+            custom_settings=self.get_additional_vagrant_config_settings()
+        )
+        return vagrant_config.substitute(
+            {'mac_address': self._random_mac()}
+        )
+
+    @staticmethod
+    def _random_mac():
+        # type: () -> str
```
Comment (schaefi) [inline kiwi/storage/subformat/vagrant_base.py]: superfluous comment

File: kiwi/storage/subformat/vagrant_base.py [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+        # type: () -> Dict[str, str]
+        return {}
+
+    @classmethod
+    def get_additional_vagrant_config_settings(cls):
+        # type: () -> str
+        return ""
+
+    def _create_box_metadata(self):
+        # type: () -> str
```
Comment (schaefi) [inline kiwi/storage/subformat/vagrant_base.py]: superfluous comment

File: kiwi/storage/subformat/vagrant_base.py [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+                os.path.basename(metadata_json),
+                os.path.basename(vagrantfile)
+            ] + [
+                os.path.basename(box_img_file)
+                for box_img_file in box_img_files
+            ]
+        )
+
+    def store_to_result(self, result):
+        # type: (Result) -> None
```
Comment (schaefi) [inline kiwi/storage/subformat/vagrant_base.py]: superfluous comment

File: kiwi/storage/subformat/vagrant_base.py [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+        if not custom_args['vagrantconfig']:
+            raise KiwiFormatSetupError(
+                'no vagrantconfig provided'
+            )
+
+        self.image_format = 'vagrant.' + self.provider + '.box'
+        self.vagrantconfig = custom_args['vagrantconfig']
+
+    def create_image_format(self):
+        # type: () -> None
```
Comment (schaefi) [inline kiwi/storage/subformat/vagrant_base.py]: superfluous comment

File: kiwi/storage/subformat/vagrant_base.py [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+from kiwi.command import Command
+from kiwi.system.result import Result  # pylint: disable=unused-import
+
+from kiwi.exceptions import (
+    KiwiFormatSetupError
+)
+
+
+def get_vagrantfile_template(custom_settings=None):
+    # type: (Optional[str]) -> Template
```
Comment (schaefi) [inline kiwi/storage/subformat/vagrant_base.py]: superfluous comment

File: kiwi/storage/subformat/vagrant_libvirt.py [hunk 11]
```diff
@@ -15,63 +15,26 @@
 # You should have received a copy of the GNU General Public License
 # along with kiwi.  If not, see <http://www.gnu.org/licenses/>
 #
-import json
 import os
-import random
+from typing import Dict  # pylint: disable=unused-import
```
Comment (schaefi) [inline kiwi/storage/subformat/vagrant_libvirt.py]: why, unused imports should be deleted

File: kiwi/storage/subformat/vagrant_libvirt.py [hunk 12]
```diff
@@ -15,63 +15,26 @@
...
 from kiwi.command import Command
 
-from kiwi.exceptions import (
-    KiwiFormatSetupError
-)
 
-
-class DiskFormatVagrantLibVirt(DiskFormatBase):
+class DiskFormatVagrantLibVirt(DiskFormatVagrantBase):
+    provider = 'libvirt'
```
Comment (schaefi) [inline kiwi/storage/subformat/vagrant_libvirt.py]: I think I overlooked something, but where is this variable used ?

File: kiwi/storage/subformat/vagrant_libvirt.py [hunk 13]
```diff
@@ -83,71 +46,26 @@ def create_image_format(self):
...
 
-    def _random_mac(self):
-        return '%02x%02x%02x%02x%02x%02x'.upper() % (
-            0x00, 0x16, 0x3e,
-            random.randrange(0, 0x7e),
-            random.randrange(0, 0xff),
-            random.randrange(0, 0xff)
-        )
+    def get_additional_vagrant_config_settings(self):
+        # type: () -> str
```
Comment (schaefi) [inline kiwi/storage/subformat/vagrant_libvirt.py]: superfluous comment

File: kiwi/storage/subformat/vagrant_libvirt.py [hunk 14]
```diff
@@ -83,71 +46,26 @@ def create_image_format(self):
...
+        """
+        Returns settings for the libvirt provider telling vagrant to use kvm.
+        """
+        return dedent('''
+        config.vm.provider :libvirt do |libvirt|
+          libvirt.driver = "kvm"
+        end
+            ''').strip()
```
Comment (schaefi) [inline kiwi/storage/subformat/vagrant_libvirt.py]: With templating in place and the fact that the string is stripped anyway this can be imho nicer indented:

```python
return dedent('''
    config.vm.provider :libvirt do |libvirt|
      libvirt.driver = "kvm"
￼    end
''').strip()
```

File: tox.ini [hunk 15]
```diff
@@ -60,8 +60,9 @@ deps = {[testenv]deps}
 changedir=test/unit
 commands =
     bash -c 'cd ../../ && ./setup.py develop'
-    pytest --no-cov-on-fail --cov=kiwi \
-        --cov-report=term-missing --cov-fail-under=100 --cov-config .coveragerc {posargs}
+    pytest --doctest-modules --no-cov-on-fail --cov=kiwi \
+        --cov-report=term-missing --cov-fail-under=100 \
+        --cov-config .coveragerc {posargs}
```
Comment (schaefi) [inline tox.ini]: Changes in tox.ini do not belong to this pull request and while I agree with the ones done here they should be done in an extra pull request to avoid confusion

File: kiwi/storage/subformat/vagrant_libvirt.py (hunk 11 above)
Comment (dcermak) [reply kiwi/storage/subformat/vagrant_libvirt.py]: That was necessary for the typing comment: `# type: () -> Dict[str, str]`.

This is an unfortunate consequence of using type hints in a python 2 compatible code base: you have to import all the types so that a type checker can evaluate these, but since you only use them in comment blocks, pylint & co. complain about unused imports since they don't evaluate type hints.

File: kiwi/storage/subformat/vagrant_libvirt.py (hunk 12 above)
Comment (dcermak) [reply kiwi/storage/subformat/vagrant_libvirt.py]: This is used in `DiskFormatVagrantBase.post_init` where the base class constructs the image format from the provider name.

File: tox.ini (hunk 15 above)
Comment (dcermak) [reply tox.ini]: Will drop them then.

File: kiwi/storage/subformat/vagrant_base.py (hunk 4 above)
Comment (dcermak) [reply kiwi/storage/subformat/vagrant_base.py]: I'll drop them.

File: kiwi/storage/subformat/vagrant_base.py (hunk 3 above)
Comment (dcermak) [reply kiwi/storage/subformat/vagrant_base.py]: Will do that!

Out of curiosity: do you know if sphinx can actually run these python code examples? I have added this as a doctest because they can be tested relatively easily via the doctest module.

File: kiwi/storage/subformat/vagrant_libvirt.py (hunk 12 above)
Comment (schaefi) [reply kiwi/storage/subformat/vagrant_libvirt.py]: Ah ok now I saw it

ok I was looking for where self.provider gets initialized. I think DiskFormatVagrantBase should initialize a default provider. The setting for ```self.image_format = 'vagrant.' + self.provider + '.box'``` is provider specific and imho doesn't belong to the base class. I think DiskFormatVagrantBase should call another post_init() such that the specialized provider class can overload that method and just set ```self.provider = 'libvirt'```. The way you did it does not loosely couple the classes and ties them hard together. The methods in the base class that references ```self.image_format``` should only reference self.provider with the value that it has at call time or its default value.

Sorry if that sounds nitpicking but it took me a while to find out how provider is used as it's class global. We should avoid that imho

File: kiwi/storage/subformat/vagrant_base.py [hunk 16]
```diff
@@ -0,0 +1,304 @@
...
+
+from kiwi.storage.subformat.base import DiskFormatBase
+from kiwi.command import Command
+
+from kiwi.exceptions import (
+    KiwiFormatSetupError
+)
+
+
+class VagrantConfigTemplate(object):
```
Comment (schaefi) [inline kiwi/storage/subformat/vagrant_base.py]: ok, nice but could you put it into an extra file below the existing template directory: ```kiwi/storage/subformat/template```

File: kiwi/storage/subformat/vagrant_base.py [hunk 17]
```diff
@@ -0,0 +1,304 @@
...
+        Vagrant.configure("2") do |config|
+    ''').strip() + os.linesep
+
+    _MAC = _INDENT + dedent('''
+        config.vm.base_mac = "${mac_address}"
+    ''').strip() + os.linesep
+
+    _END = dedent('''
+        end
+    ''').strip()
```
Comment (schaefi) [inline kiwi/storage/subformat/vagrant_base.py]: I usually avoid class global values especially when we create an instance object of the class. The template is created in code later as ```template = VagrantConfigTemplate()``` and we operate on template. Thus all this values from above _INDENT, _HEADER, etc... deserves a constructor and should imho be set in the __init__ method.

File: kiwi/storage/subformat/vagrant_base.py [hunk 18]
```diff
@@ -0,0 +1,304 @@
...
+
+    @classmethod
+    def get_template(cls, custom_settings=None):
```
Comment (schaefi) [inline kiwi/storage/subformat/vagrant_base.py]: I think this should be a normal method not a classmethod because we create an object from VagrantConfigTemplate

**Comment by schaefi** [comment]
Also please check on the codacy report. I already marked the random issue as to be ignore because we don't really use it for security reasons as they complain :) The other issue however is valid and should be fixed. Thanks

File: kiwi/storage/subformat/vagrant_base.py (hunk 17 above)
Comment (dcermak) [reply kiwi/storage/subformat/vagrant_base.py]: I'll change that.

I have used class global values since they will be constructed once the module is imported and not every time a template is created. But since it's unlikely that we will create two templates, it won't make a difference.

File: kiwi/storage/subformat/vagrant_base.py (hunk 18 above)
Comment (dcermak) [reply kiwi/storage/subformat/vagrant_base.py]: Yes, I'll change that back to an ordinary method. I used a classmethod since codacy would have complained about `get_template` not using self.

**Comment by dcermak** [comment]
I can't reply to this comment directly, so I'll do it here:

> Ah ok now I saw it
>
> ok I was looking for where self.provider gets initialized. I think DiskFormatVagrantBase should initialize a default provider. The setting for self.image_format = 'vagrant.' + self.provider + '.box' is provider specific and imho doesn't belong to the base class. I think DiskFormatVagrantBase should call another post_init() such that the specialized provider class can overload that method and just set self.provider = 'libvirt'. The way you did it does not loosely couple the classes and ties them hard together. The methods in the base class that references self.image_format should only reference self.provider with the value that it has at call time or its default value.
>
> Sorry if that sounds nitpicking but it took me a while to find out how provider is used as it's class global. We should avoid that imho

I have just implemented this and it has a disadvantage imho: I need a second `post_init` style function (which I have called `vagrant_post_init`). Using `post_init` and then calling the base classes' `post_init` via `super()` does not work in Python2.7, where this always results in `DiskFormatBase.post_init` being called (which does nothing).

I have just pushed this change, so that you can judge for yourself whether you like it or not.

**Comment by schaefi** [comment]
Looks good to me now. Thanks. I'll merge and see if Virt:App staging builds still succeeds
